contrib/pgo-lto/Makefile

```diff
@@ -0,0 +1,57 @@
+.PHONY: top clean
+
+STAGE1_BUILD:=$(CURDIR)/stage1.build
+STAGE2_BUILD:=$(CURDIR)/stage2.build
+
+PROFILE_DIR:=$(CURDIR)/profiles
+PROFILE_FILE:=$(PROFILE_DIR)/merged.prof
+JULIA_ROOT:=$(CURDIR)/../..
+CXXFILT:=c++filt
+
+# Note: can't use BB's LLVM right now; -fprofile-generate compiles, but doesn't
+# generate profiles at runtime
+LLVM_PROFDATA:=llvm-profdata
+export CC:=clang
+export CXX:=clang++
+export LD:=ld.lld
+export AR:=llvm-ar
+export RANLIB:=llvm-ranlib
+export USECLANG:=1
+export USE_BINARYBUILDER_LLVM:=0
+# export VERBOSE:=1
+
+# When building a single libLLVM.so we need to increase -vp-counters-per-site
+# significantly
+COUNTERS_PER_SITE:=6
+
+AFTER_STAGE1_MESSAGE:=Consider removing $(PROFILE_DIR) to start with a clean slate. $\
+    Then run Julia to collect realistic profile data, for example: \`julia -O3 -e $\
+    'using Pkg; Pkg.add("Unitful"); Pkg.test("Unitful")'\`. This should produce about $\
+	15MB of data in $(PROFILE_DIR). Note that running extensive scripts may result in $\
+	counter overflows, which can be detected by running \`make $(PROFILE_FILE)\`. $\
+	Afterwards run `make stage2`.
+
+$(STAGE1_BUILD) $(STAGE2_BUILD):
+	$(MAKE) -C $(JULIA_ROOT) O=$@ configure
+
+stage1: export CFLAGS=-fprofile-generate=$(PROFILE_DIR) -Xclang -mllvm -Xclang -vp-counters-per-site=$(COUNTERS_PER_SITE)
+stage1: export CXXFLAGS=-fprofile-generate=$(PROFILE_DIR) -Xclang -mllvm -Xclang -vp-counters-per-site=$(COUNTERS_PER_SITE)
+stage1: export LDFLAGS=-fuse-ld=lld -flto=thin -fprofile-generate=$(PROFILE_DIR)
```

Comment:
I actually didn't know you could attach environment variables to targets like this in `Make`!  This is very cool!

Quick demonstration to anyone else watching, who wants to understand better how this interacts with rules and dependencies:

```Makefile
$ cat Makefile 
all: foo bar foobar

# This rule will have `$FOO` defined within it
foo:
        @echo "[foo]    FOO: $${FOO}"

# This rule will not
bar:
        @echo "[bar]    FOO: $${FOO}"

# Even though this rule depends on `foo`, it won't have `$FOO` defined.
foobar: foo bar
        @echo "[foobar] FOO: $${FOO}"

# Attach an environment variable to `foo`
foo: export FOO=foo
```

```
$ make
[foo]    FOO: foo
[bar]    FOO: 
[foobar] FOO: 
```

Reply:
Yeah, this is neat! :D